File: Python/ByRound/1869/1869_A_Make_It_Zero.py
# Problem: CF 1869 A - Make It Zero
# https://codeforces.com/contest/1869/problem/A

"""
Code Purpose:
This solution addresses the problem of making all elements of an array equal to zero using at most 8 operations. 
Each operation selects a subarray and replaces all its elements with the XOR of the elements in that subarray.
The approach is based on the observation that:
- If the array length n is even, two operations suffice: one to XOR the entire array, and another to XOR the entire array again.
- If the array length n is odd, we use 4 operations:
    1. XOR from index 1 to n-1
    2. XOR from index 1 to n-1 again
    3. XOR from index 2 to n
    4. XOR from index 2 to n again

Algorithms/Techniques:
- Bitwise XOR operation
- Observation-based greedy approach

Time Complexity: O(n) per test case, where n is the length of the array.
Space Complexity: O(1) excluding the input storage.
"""

from sys import stdin

input = stdin.readline
for _ in range(int(input())):
    n = int(input())
    b = input()
    if n % 2:
        print(4)  # For odd n, 4 operations are needed
        print(1, n - 1)  # First operation on all elements except last
        print(1, n - 1)  # Second operation on all elements except last
        print(2, n)      # Third operation on all elements except first
        print(2, n)      # Fourth operation on all elements except first
    else:
        print(2)  # For even n, 2 operations are sufficient
        print(1, n)  # Operation on entire array
        print(1, n)  # Operation on entire array again


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/A/1869_A_Make_It_Zero.cpp
// Problem: CF 1869 A - Make It Zero
// https://codeforces.com/contest/1869/problem/A

/*
 * Problem: Make It Zero
 * Algorithm: Greedy approach based on array length parity
 * Time Complexity: O(n) per test case, where n is the length of the array
 * Space Complexity: O(1), only using a constant amount of extra space
 *
 * The strategy is:
 * - If the array length n is odd, we perform 4 operations:
 *   1. Operation on [2, n]: XOR the subarray from index 2 to n
 *   2. Operation on [2, n]: XOR the same subarray again
 *   3. Operation on [1, 2]: XOR the first two elements
 *   4. Operation on [1, 2]: XOR the first two elements again
 * - If the array length n is even, we perform 2 operations:
 *   1. Operation on [1, n]: XOR the entire array
 *   2. Operation on [1, n]: XOR the entire array again
 *
 * This ensures all elements become 0, as XOR of all elements with themselves
 * results in 0, and we are effectively "zeroing out" the array using these
 * operations.
 */

#include <cstdio>
int main() {
  long t;
  scanf("%ld", &t);
  while (t--) {
    long n;
    scanf("%ld", &n);
    for (long p = 0; p < n; p++) {
      long x;
      scanf("%ld", &x);
    }
    if (n % 2) {
      printf("4\n2 %ld\n2 %ld\n1 2\n1 2\n", n, n);
    } else {
      printf("2\n1 %ld\n1 %ld\n", n, n);
    }
  }
}


// https://github.com/VaHiX/CodeForces/